examples/llm/src/models/layers/attention.py

```diff
@@ -12,377 +13,343 @@
-            attn_mask = attn_mask.expand(batch_size, heads, seq_len,
-                                         seq_len).reshape(-1, seq_len, seq_len)
-
-        return attn_mask
-
-
-class FlashCausalAttention(nn.Module):
-
-    def __init__(self, cfg: DictConfig, device: Optional[str] = None):
-        super().__init__()
-        try:
-            from flash_attn.flash_attention import (  # type: ignore
-                FlashAttention, FlashMHA)
-        except ImportError as e:
-            raise e
-
-        self.clip_qkv = cfg.get('attn_clip_qkv')
-        self.attn_qk_ln = cfg.get('attn_qk_ln')
-        self.softmax_scale = cfg.get('softmax_scale')
-        self.d_model = cfg.d_model
-        self.n_heads = cfg.n_heads
-
-        if self.attn_qk_ln or self.clip_qkv or self.softmax_scale:
-            self.W_qkv = nn.Linear(self.d_model,
-                                   3 * self.d_model,
-                                   bias=True,
-                                   device=device)
-            self.inner_attn = FlashAttention(attention_dropout=cfg.attn_pdrop,
-                                             device=device,
-                                             softmax_scale=self.softmax_scale)
-            self.out_proj = nn.Linear(self.d_model,
-                                      self.d_model,
-                                      bias=True,
-                                      device=device)
-            # for param init fn
-            fuse_splits = (cfg.d_model, 2 * cfg.d_model)
-            self.W_qkv._fused = (0, fuse_splits)  # type: ignore
-            self.out_proj._is_residual = True  # type: ignore
-
-            if self.attn_qk_ln:
-                layernorm_class = nn.LayerNorm if not cfg.get(
-                    'low_precision_layernorm', False) else LPLayerNorm
-                self.q_ln = layernorm_class(self.d_model, device=device)
-                self.k_ln = layernorm_class(self.d_model, device=device)
-        else:
-            self.mhsa = FlashMHA(
-                embed_dim=cfg.d_model,
-                num_heads=cfg.n_heads,
-                attention_dropout=cfg.attn_pdrop,
-                bias=True,
-                batch_first=True,
-                causal=True,
-                device=device,
+from torch import nn
+
+
+def scaled_multihead_dot_product_attention(
+    query,
+    key,
+    value,
+    n_heads,
+    softmax_scale=None,
+    attn_bias=None,
+    key_padding_mask=None,
+    is_causal=False,
+    dropout_p=0.0,
+    training=False,
+    needs_weights=False,
+):
+    q = rearrange(query, 'b s (h d) -> b h s d', h=n_heads)
+    k = rearrange(key, 'b s (h d) -> b h d s', h=n_heads)  # includes key.t()
+    v = rearrange(value, 'b s (h d) -> b h s d', h=n_heads)
+
+    b, _, s_q, d = q.shape
+    s_k = k.size(-1)
+
+    if softmax_scale is None:
+        softmax_scale = 1 / math.sqrt(d)
+
+    attn_weight = q.matmul(k)
+    attn_weight *= softmax_scale
```

Comment:
```suggestion
    attn_weight = q.matmul(k) * softmax_scale
```

I have a deep fear of using += or *= style operations on torch tensors. Long ago there were autograd issues with doing so.

Reply:
My way: attn_weight is generated (`attn_weight = q.matmul(k)`), attn_weight is in-place modified (`attn_weight *= softmax_scale`)

Your way: `q.matmul(k)` instantiates `attn_weight`, ` * softmax_scale` is NOT an in-place mul of `attn_weight` so another set of attn_weights are generated...

I'll do:
```suggestion
    attn_weight = q.matmul(k)
    attn_weight.mul_(softmax_scale)
```
to avoid using += or *= style operations on torch tensors.

Reply:
Actually let us not make them in-place. In-place operations don't interact nicely with other components including torch.compile. Torch.compile will anyway fuse these so keeping them separate is fine.

Reply:
I'll nuke a lot of in-place ops then